Fix 584458794

Automated tests found a bug in the directory: https://github.com/foundationdb-rs/foundationdb-rs/runs/5849103291?check_suite_focus=true

```
 + ./bindings/bindingtester/bindingtester.py --num-ops 1000 --api-version 630 --test-name directory --concurrency 1 rust --no-directory-snapshot-ops --compare python

Creating test at API version 630
Generating directory test at seed 584458794 with 1000 op(s) and 1 concurrent tester(s)...

Inserting test into database...

Running tester '/home/runner/work/foundationdb-rs/foundationdb-rs/target/debug/bindingtester test_spec 630'...

Reading results from '('tester_output', 'stack')'...
Reading results from '('tester_output', 'directory_log', 'directory')'...
Reading results from '('tester_output', 'directory_log', 'subspace')'...

Inserting test into database...

Running tester 'python /home/runner/work/foundationdb-rs/foundationdb-rs/target/foundationdb_build/foundationdb/bindings/bindingtester/../python/tests/tester.py test_spec 630'...

Reading results from '('tester_output', 'stack')'...
Reading results from '('tester_output', 'directory_log', 'directory')'...
Reading results from '('tester_output', 'directory_log', 'subspace')'...

Comparing results from '('tester_output', 'stack')'...

Incorrect result: 
  rust   - ('tester_output', 'stack', 184, 1684) = b'\x01RESULT_NOT_PRESENT\x00'
  python - ('tester_output', 'stack', 184, 1684) = b'\x01\x01ERROR\x00\xff\x011020\x00\xff\x00'

Incorrect result: 
  rust   - None
  python - ('tester_output', 'stack', 185, 1693) = b'\x01DIRECTORY_ERROR\x00'

Incorrect result: 
  rust   - ('tester_output', 'stack', 204, 1832) = b'\x01DIRECTORY_ERROR\x00'
  python - None

Incorrect result: 
  rust   - ('tester_output', 'stack', 206, 1842) = b'\x01DIRECTORY_ERROR\x00'
  python - None

Comparing results from '('tester_output', 'directory_log', 'directory')'...

Incorrect result: 
  rust   - ('tester_output', 'directory_log', 'directory', 5, 'children') = b'\x021\x00'
  python - ('tester_output', 'directory_log', 'directory', 5, 'children') = b''

Incorrect result: 
  rust   - ('tester_output', 'directory_log', 'directory', 7, 'children') = b'\x021\x00'
  python - ('tester_output', 'directory_log', 'directory', 7, 'children') = b''

Incorrect result: 
  rust   - ('tester_output', 'directory_log', 'directory', 130, 'children') = b'\x021\x00'
  python - ('tester_output', 'directory_log', 'directory', 130, 'children') = b''

Incorrect result: 
  rust   - ('tester_output', 'directory_log', 'directory', 131, 'children') = b''
  python - None

Incorrect result: 
  rust   - ('tester_output', 'directory_log', 'directory', 131, 'exists') = b'\x15\x01'
  python - None

Incorrect result: 
  rust   - ('tester_output', 'directory_log', 'directory', 131, 'layer') = b'\x01\x00'
  python - None

Incorrect result: 
  rust   - ('tester_output', 'directory_log', 'directory', 131, 'path') = b'\x02default1\x00'
  python - None

Incorrect result: 
  rust   - None
  python - ('tester_output', 'directory_log', 'directory', 143, 'children') = b''

Incorrect result: 
  rust   - None
  python - ('tester_output', 'directory_log', 'directory', 143, 'exists') = b'\x15\x01'

Incorrect result: 
  rust   - None
  python - ('tester_output', 'directory_log', 'directory', 143, 'layer') = b'\x01\x00'

Incorrect result: 
  rust   - None
  python - ('tester_output', 'directory_log', 'directory', 143, 'path') = b'\x021\x00'

Incorrect result: 
  rust   - None
  python - ('tester_output', 'directory_log', 'directory', 144, 'children') = b''

Incorrect result: 
  rust   - None
  python - ('tester_output', 'directory_log', 'directory', 144, 'exists') = b'\x15\x01'

Incorrect result: 
  rust   - None
  python - ('tester_output', 'directory_log', 'directory', 144, 'layer') = b'\x01\x00'

Incorrect result: 
  rust   - None
  python - ('tester_output', 'directory_log', 'directory', 144, 'path') = b'\x02default1\x00'

Comparing results from '('tester_output', 'directory_log', 'subspace')'...

Test with seed 584458794 and concurrency 1 had 19 incorrect result(s) and 0 error(s) at API version 630
Completed directory test with random seed 584458794 and 1000 operations

Error: Process completed with exit code 1.
```

--- a/foundationdb/src/database.rs
+++ b/foundationdb/src/database.rs
@@ -240,7 +240,7 @@ mod boxed {
         ) -> Pin<Box<dyn Future<Output = Result<T, E>> + Send + 'a>>,
         E: TransactError,
     {
-        let r = (&mut f.f)(&trx, &mut f.d).await;
+        let r = (f.f)(&trx, &mut f.d).await;
         (f, trx, r)
     }
 
@@ -293,7 +293,7 @@ mod boxed_local {
         ) -> Pin<Box<dyn Future<Output = Result<T, E>> + 'a>>,
         E: TransactError,
     {
-        let r = (&mut f.f)(&trx, &mut f.d).await;
+        let r = (f.f)(&trx, &mut f.d).await;
         (f, trx, r)
     }
 

--- a/foundationdb/src/directory/directory_layer.rs
+++ b/foundationdb/src/directory/directory_layer.rs
@@ -550,7 +550,7 @@ impl DirectoryLayer {
             };
         }
 
-        Ok(node.list_sub_folders(trx).await?)
+        node.list_sub_folders(trx).await
     }
 
     async fn move_to_internal(

--- a/foundationdb/src/directory/directory_layer.rs
+++ b/foundationdb/src/directory/directory_layer.rs
@@ -308,7 +308,7 @@ impl DirectoryLayer {
 
         let layer = layer.unwrap_or_default();
 
-        self.check_version(trx, allow_create).await?;
+        self.check_version(trx, true).await?;
         let new_prefix = self.get_prefix(trx, prefix).await?;
 
         let is_prefix_free = self
@@ -567,35 +567,42 @@ impl DirectoryLayer {
             return Err(DirectoryError::CannotMoveBetweenSubdirectory);
         }
 
-        let old_node = self
-            .find(trx, old_path)
-            .await?
-            .ok_or(DirectoryError::PathDoesNotExists)?;
+        let old_node = self.find(trx, old_path).await?;
+        let new_node = self.find(trx, new_path).await?;
+
+        let old_node = match old_node {
+            None => return Err(DirectoryError::PathDoesNotExists),
+            Some(n) => n,
+        };
         let old_node_exists_in_partition = old_node.is_in_partition(false);
 
-        if let Some(new_node) = self.find(trx, new_path).await? {
-            let new_node_exists_in_partition = new_node.is_in_partition(false);
-            if old_node_exists_in_partition || new_node_exists_in_partition {
-                if !old_node_exists_in_partition
-                    || !new_node_exists_in_partition
-                    || !old_node.current_path.eq(&new_node.current_path)
-                {
+        match new_node {
+            None => {
+                if old_node_exists_in_partition {
                     return Err(DirectoryError::CannotMoveBetweenPartition);
                 }
-
-                return new_node
-                    .get_contents()?
-                    .move_to(
-                        trx,
-                        &old_node.get_partition_subpath(),
-                        &new_node.get_partition_subpath(),
-                    )
-                    .await;
             }
+            Some(new_node) => {
+                let new_node_exists_in_partition = new_node.is_in_partition(false);
+                if old_node_exists_in_partition || new_node_exists_in_partition {
+                    if !old_node_exists_in_partition
+                        || !new_node_exists_in_partition
+                        || !old_node.current_path.eq(&new_node.current_path)
+                    {
+                        return Err(DirectoryError::CannotMoveBetweenPartition);
+                    }
 
-            return Err(DirectoryError::DirAlreadyExists);
-        } else if old_node_exists_in_partition {
-            return Err(DirectoryError::CannotMoveBetweenPartition);
+                    return new_node
+                        .get_contents()?
+                        .move_to(
+                            trx,
+                            &old_node.get_partition_subpath(),
+                            &new_node.get_partition_subpath(),
+                        )
+                        .await;
+                }
+                return Err(DirectoryError::DirAlreadyExists);
+            }
         }
 
         let (new_path_last, parent_path) = new_path

--- a/scripts/run_bindingtester.sh
+++ b/scripts/run_bindingtester.sh
@@ -22,6 +22,9 @@ cd "${fdb_builddir:?}/foundationdb"
 ./bindings/bindingtester/bindingtester.py --num-ops 1000 --api-version 630 --test-name directory --concurrency 1 rust --no-directory-snapshot-ops --compare python --seed 1951034301
 ./bindings/bindingtester/bindingtester.py --num-ops 1000 --api-version 630 --test-name directory --concurrency 1 rust --no-directory-snapshot-ops --compare python --seed 1700644945
 
+# https://github.com/foundationdb-rs/foundationdb-rs/issues/42
+./bindings/bindingtester/bindingtester.py --num-ops 1000 --api-version 630 --test-name directory --concurrency 1 rust --no-directory-snapshot-ops --compare python --seed 584458794
+
 START=1
 END=${1-1}
 for i in $(eval echo "{$START..$END}")

